runtime/src/util/batch_return.rs

```diff
@@ -101,39 +101,47 @@ impl fmt::Display for BatchReturn {
 /// indexed against only those successful items.
 /// E.g. stack([OK, E1, OK, E2], [OK, E3], [E4]) => [E4, E1, E3, E2]
 pub fn stack(batch_returns: &[BatchReturn]) -> BatchReturn {
-    if batch_returns.is_empty() {
-        return BatchReturn::empty();
-    }
-    let mut base = batch_returns[0].clone();
-    for nxt in &batch_returns[1..] {
+    fn merge_layer<'a>(
```

Comment:
Yes, this is recursive. But the recursion should be short and it lets us avoid allocations.